content/post/simulating-event-windows.md

```diff
@@ -0,0 +1,171 @@
+64.242.88.10 - - [07/Mar/2004:16:32:50 -0800] "GET /twiki/bin/view/Main/WebChanges HTTP/1.1" 200 40520
+```
+
+we will expect an output as follows
+
+```
+2004-03-07 200 3
+2004-03-07 401 2
+```
+
+We will also adopt a small trick. Because we are simulating event-based windowing, we will introduce a "token" marker after each day, that our topology will have to interpret correctly. For example
+
+```
+64.242.88.10 - - [07/Mar/2004:16:24:16 -0800] "GET /twiki/bin/view/Main/ABlogPage HTTP/1.1" 200 4924
+64.242.88.10 - - [07/Mar/2004:16:29:16 -0800] "GET /twiki/bin/edit/Main/Header_checks?topicparent=Main.ConfigurationVariables HTTP/1.1" 401 12851
+64.242.88.10 - - [07/Mar/2004:16:30:29 -0800] "GET /twiki/bin/attach/Main/OfficeLocations HTTP/1.1" 401 12851
+END_OF_DAY
+64.242.88.10 - - [08/Mar/2004:06:11:48 -0800] "GET /twiki/bin/view/TWiki/AnotherPage HTTP/1.1" 200 3732
+64.242.88.10 - - [08/Mar/2004:06:12:50 -0800] "GET /twiki/bin/view/Main/YetAnotherPage HTTP/1.1" 200 40520
+```
+
+### Topology
+
+The simplest possible topology to do this would be composed of three elements, according to the following diagram
+
+![High Level Logfiles Diagram](/images/post/simulate-event-windowing-wallaroo/topology.png)
+
+A fully working version of the code from this post can be found in the [examples](http://github.com/) directory of our repository, together with instructions on how to build it and run it on some example data.
+
+We set up our Wallaroo topology as follows
+
+```
+def application_setup(args):
+    in_host, in_port = wallaroo.tcp_parse_input_addrs(args)[0]
+    out_host, out_port = wallaroo.tcp_parse_output_addrs(args)[0]
+
+    ab = wallaroo.ApplicationBuilder("Apache Log file analysis")
+    ab.new_pipeline("convert",
+                    wallaroo.TCPSourceConfig(in_host, in_port, Decoder()))
+    ab.to(Count)
+    ab.to_sink(wallaroo.TCPSinkConfig(out_host, out_port, Encoder()))
+    return ab.build()
+```
+
+#### Decoder
+
+Our decoder needs to be able to correctly identify the `END_OF_DAY` message, and translate that into an appropriate message that will signal to the computations that they need to roll over to the next day. We will call this a `BoundaryMessage`, and all other messages will be `LogLine`s.
+
+```
+class Decoder(object):
+    def header_length(self):
+        return 4
+
+    def payload_length(self, bs):
+        return struct.unpack(">I", bs)[0]
+
+    def decode(self, bs):
+        if bs == 'END_OF_DAY':
+            return BoundaryMessage()
+        else:
+            return LogLine(bs)
+```
+
+#### Computation
+
+The computation itself needs to be able to do the following things:
+
+ - On the first `LogLine` for each window, determine which day we are counting for
+ - On all `LogLine`s, accumulate the counts
+ - On receipt of a `BoundaryMessage`, send all the output required as a
+   `SummaryMessage`
+
+```
+class Count(object):
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
```

Comment:
This different than using `to_stateful`, which is how all of our current examples interact with state. How come?